Enable changed var with ufw check mode

jbarotin opened the pull request:
##### SUMMARY
<!--- Describe the change below, including rationale and design decisions -->
This is a proposal in order to have boolean "changed" value while activate the check mode of the ufw module. 
Fixes #45604 

<!--- HINT: Include "Fixes #nnn" if you are fixing an existing issue -->

##### ISSUE TYPE
<!--- Pick one below and delete the rest -->
- Feature Pull Request

##### COMPONENT NAME
ufw 

##### ADDITIONAL INFORMATION
<!--- Include additional information to help people understand the change here -->
<!--- A step-by-step reproduction of the problem is helpful if there is no related issue -->

<!--- Paste verbatim command output below, e.g. before and after your change -->
```paste below

```

jbarotin committed:
Enable 'changed' var with ufw check mode

jbarotin commented:
+label needs_ci

jbarotin commented:
I don't understand why CI test fail (time out issue). Is it possible to run them again ?

felixfontein commented:
The current test failure seems to be unrelated to this PR. This (unfortunately) happens from time to time, since a lot of tests depend on online resources being available (to install packages etc.). You can get this individual failed test restarted by asking on IRC (#ansible-devel); a core team member can restart it (most effective during work-week, when people are around :) ). Otherwise, you can simply close and re-open the PR, this will trigger a completely new CI run.

felixfontein commented on lib/ansible/modules/system/ufw.py:
```diff
@@ -245,10 +245,11 @@ def execute(cmd):
 
         cmds.append(cmd)
-        (rc, out, err) = module.run_command(cmd)
+        (rc, out, err) = module.run_command(cmd, environ_update={"LANG": "en_EN"})
```
Do you mean `en_US`? I don't think the locale `en_EN` actually exists.

jbarotin replied:
You're right, my current locale is fr_FR I did a quick and wrong transposition to en. I'll change that to en_US in my next commit.

felixfontein commented on lib/ansible/modules/system/ufw.py:
```diff
@@ -296,20 +297,36 @@ def ufw_version():
+                else:
+                    (_, ufw_state, _) = execute([[ufw_bin], ["status"]])
+                    ufw_enabled = ufw_state.find("active") != -1
+                    if (value == 'disabled' and ufw_enabled) or (value == 'enabled' and ufw_enabled):
```
There's a `not` missing somewhere in here :)
```suggestion
                    if (value == 'disabled' and ufw_enabled) or (value == 'enabled' and not ufw_enabled):
```

felixfontein commented on lib/ansible/modules/system/ufw.py:
```diff
@@ -296,20 +297,36 @@ def ufw_version():
         elif command == 'logging':
-            execute(cmd + [[command], [value]])
+            if module.check_mode:
+                changed = True
```
You can find the current logging level with `ufw status verbose`. For me, it contains the line `Logging: on (low)`. This should allow to determine `changed`.

felixfontein commented on lib/ansible/modules/system/ufw.py:
```diff
@@ -296,20 +297,36 @@ def ufw_version():
         elif command == 'default':
-            execute(cmd + [[command], [value], [params['direction']]])
+            if module.check_mode:
+                changed = True
```
Same here. (`Default: deny (incoming), allow (outgoing), deny (routed)`)

jbarotin replied:
I don't know how to interpret this line. What does the second deny mean ? There's nothing in man about that.

felixfontein replied:
From the manual:
```ufw [--dry-run] default allow|deny|reject [incoming|outgoing|routed]

[...]
       default allow|deny|reject DIRECTION
              change  the  default  policy  for traffic going DIRECTION, where
              DIRECTION is one of incoming,  outgoing  or  routed.  Note  that
              existing  rules  will have to be migrated manually when changing
              the default policy. See RULE SYNTAX for more on deny and reject.
```
It means that both "directions" `incoming` and `routed` are set to `deny`, and the direction `outgoing` is set to `allow`.

felixfontein commented on lib/ansible/modules/system/ufw.py:
```diff
@@ -336,14 +353,19 @@ def ufw_version():
+    if module.check_mode:
+        return module.exit_json(changed=changed, commands=cmds)
+    else:
+        (_, post_state, _) = module.run_command(ufw_bin + ' status verbose')
```
While you're at it, could you change `ufw_bin + ' status verbose'` to `[ufw_bin, 'status', 'verbose']`?

felixfontein commented on lib/ansible/modules/system/ufw.py:
```diff
@@ -336,14 +353,19 @@ def ufw_version():
+        return module.exit_json(changed=changed, commands=cmds)
+    else:
+        (_, post_state, _) = module.run_command(ufw_bin + ' status verbose')
+        (_, post_rules, _) = module.run_command("grep '^### tuple' /lib/ufw/user.rules /lib/ufw/user6.rules /etc/ufw/user.rules /etc/ufw/user6.rules")
```
Same here, use the array form (i.e. `['grep', '^### tuple', '/lib/...`). (And everywhere else where it appears.)

Also, you might want to replace `'grep'` by the result of `module.get_bin_path('grep', True)` (which should also be stored in a variable, similarly to `ufw_bin`, and used everywhere where `grep` is used directly).

felixfontein commented on lib/ansible/modules/system/ufw.py:
```diff
@@ -336,14 +353,19 @@ def ufw_version():
-            execute(cmd)
+            (_, rules_dry, _) = execute(cmd)
+            if module.check_mode:
+                if rules_dry.find("Rules updated") != -1:
```
Doesn't work for me when deleting non-existing rules: the output of `--dry-run` will always contain `Rules updated` for me. (It does work for adding rules, though.)

jbarotin replied:
That's silly, It works for me :)

I'm using Ubuntu 18.04 LTS and UFW version 0.35. I tested my code with the "Rules updated" check on this version. I have also checked on Ubuntu 16.04 LTS and it's the same version of UFW. 

What is your OS and UFW version ?

jbarotin replied:
To complete here it is the output of commands on my dev computer (Ubuntu 18.04)

#LANG=en_US ufw status verbose
```
Status: active
Logging: on (low)
Default: deny (incoming), allow (outgoing), deny (routed)
New profiles: skip

To                         Action      From
--                         ------      ----
7000/tcp                   ALLOW IN    Anywhere                  
7000/tcp (v6)              ALLOW IN    Anywhere (v6)             

7000                       ALLOW FWD   Anywhere                  
7000 (v6)                  ALLOW FWD   Anywhere (v6)   
```

#LANG=en_US ufw --dry-run route allow 7000 
```
Skipping adding existing rule
Skipping adding existing rule (v6)
```

felixfontein replied:
I'm also using `ufw 0.35`, both with Arch Linux and Debian 9 (stretch).

I also get the `Skipping adding existing rule` message when trying to add a rule which is already there. I don't get such a message when *removing* a rule that doesn't exist. The information `Could not delete non-existent rule` is only printed when run *without* `--dry-run` :(

jbarotin replied:
Ok. I understand. In this case, I'll compare content of /etc/user.rules & /etc/user6.rule with the result of `ufw --dry-run delete <my rule>`

felixfontein replied:
Yes, that's probably the best way to proceed...

jbarotin commented:
@felixfontein thanks a lot for the review, that's my first PR on project big as Ansible.  I'll correct all your pertinent comment this in a future commit.

Do you know a way to execute CI tests on my computer without sending PR ?

felixfontein commented:
There's a lot of material on the docs site:
https://docs.ansible.com/ansible/latest/dev_guide/testing_sanity.html
https://docs.ansible.com/ansible/latest/dev_guide/testing_units_modules.html
https://docs.ansible.com/ansible/latest/dev_guide/testing_integration.html

Since there are no unit and integration tests for ufw, it suffices to run the sanity tests. The most important (if you don't modify documentation) is usually `flake8` (the test is called `pep8` in Ansible), which you can also have easier by `pip install flake8` and then running `flake8 path/to/ufw.py`.

jbarotin committed:
Fix from comment of the PR + Unit Test

felixfontein commented on lib/ansible/modules/system/ufw.py:
```diff
@@ -245,12 +245,17 @@ def execute(cmd):
         return (rc, out, err)
 
+    def get_current_rules():
+        return execute([[grep_bin], ["'^### tuple'"], ["/lib/ufw/user.rules"],
```
There's no need to add additional quotes. Simply replace `"'^### tuple'"` by `'^### tuple'`.

jbarotin replied:
Not sure that work, without quote, It send me the following error message `/bin/grep: tuple: No such file or directory`

felixfontein replied:
See my new comment; that should fix this problem.

felixfontein commented on lib/ansible/modules/system/ufw.py:
```diff
@@ -292,12 +297,16 @@ def ufw_version():
-    (_, pre_state, _) = module.run_command(ufw_bin + ' status verbose')
-    (_, pre_rules, _) = module.run_command("grep '^### tuple' /lib/ufw/user.rules /lib/ufw/user6.rules /etc/ufw/user.rules /etc/ufw/user6.rules")
+    (_, pre_state, _) = execute([[ufw_bin], ['status'], ['verbose']])
+    (_, pre_rules, _) = get_current_rules()
```
Why not return just the middle part of the tuple (instead of the full tuple)?

jbarotin replied:
Yes we can make it simpler.

felixfontein commented on lib/ansible/modules/system/ufw.py:
```diff
@@ -309,22 +318,41 @@ def ufw_version():
+                extract = re.search(r'Logging: ([onf]+) \(([a-z]+)\)', pre_state)
+                if extract:
+                    current_level = extract.group(1)
+                    current_on_off_value = extract.group(2)
```
Isn't the first group `on`/`off`, and the second group the level?

felixfontein commented on lib/ansible/modules/system/ufw.py:
```diff
@@ -309,22 +318,41 @@ def ufw_version():
         elif command == 'logging':
             if module.check_mode:
-                changed = True
+                extract = re.search(r'Logging: ([onf]+) \(([a-z]+)\)', pre_state)
```
Why not `(on|off)` instead of `([onf]+)`?

felixfontein commented on lib/ansible/modules/system/ufw.py:
```diff
@@ -309,22 +318,41 @@ def ufw_version():
         elif command == 'default':
             if module.check_mode:
-                changed = True
+                extract = re.search(r'Default: ([denyalow]+) \(incoming\), ([denyalow]+) \(outgoing\), ([denyallow]+) \(routed\)', pre_state)
```
Why not `(deny|allow)` instead of `([denyalow]+)`?

felixfontein commented on lib/ansible/modules/system/ufw.py:
```diff
@@ -309,22 +318,41 @@ def ufw_version():
-                changed = True
+                extract = re.search(r'Default: ([denyalow]+) \(incoming\), ([denyalow]+) \(outgoing\), ([denyallow]+) \(routed\)', pre_state)
+                if extract:
+                    defaults = {}
```
I would rather say `current_values` instead of `defaults`, since these are not the default values, but the current values.

jbarotin replied:
That's refer to the firewall default rules, I'll rename to `current_default_values`.

felixfontein replied:
Ah. Yes, good idea; the new name is much better and won't confuse me that much ;)

felixfontein commented on lib/ansible/modules/system/ufw.py:
```diff
@@ -353,17 +381,23 @@ def ufw_version():
 
+            # delete rules with --dry-run doesn't display "Rules updated" messages
+            # we compare ## tupple
+            cmd.append([module.check_mode and module.boolean(params['delete']), "| grep -E '^### tuple'"])
```
This won't work. The command is not executed by a shell, so `ufw` will get the command line argument `"| grep -E '^### tuple'"` (as a single argument value).

jbarotin replied:
Have you got any idea how to simulate a grep in python or with ansible lib ?

felixfontein replied:
`output = ''.join([line for line in output.splitlines(keepends=True) where line.startswith('### tuple')])`

That will split the string `output` into lines (keeping the line endings), take only the lines which start with `### tuple`, and concatenate these lines together.

felixfontein commented on lib/ansible/modules/system/ufw.py:
```diff
@@ -292,12 +297,16 @@ def ufw_version():
 
     changed = False
+
+    debug = ""
```
What do you need this for?

jbarotin replied:
Oops, I forget to remove that temporary var.

jbarotin committed:
Fix on ufw module after the second review

- delete rules change works in check mode
- simplify execute def & use it on every call process
- improved regexp
- rename vars defaults to current_default_values

felixfontein commented on lib/ansible/modules/system/ufw.py:
```diff
@@ -245,10 +245,16 @@ def execute(cmd):
         cmd = ' '.join(map(itemgetter(-1), filter(itemgetter(0), cmd)))
```
Ok, I really didn't expect the code to be *that* crazy. Could you replace this by
```suggestion
        cmd = list(map(itemgetter(-1), filter(itemgetter(0), cmd)))
```
? `module.run_command()` is totally capable of processing a list (instead of a string), and in fact, that's the WAY better method to call it. If you pass a string, that string first has to be parsed again (to be split as a list), and that list is then passed on...

jbarotin replied:
All right let's try :)

jbarotin replied:
I think this will change a lot the comportment of this function. There is commands with space mixed with command in list.

I haven't time to execute full regression test on this module (and there is no unit test). For a first version, and to enable check mode, we should keep the quote in the `grep '^### tuple'`

felixfontein replied:
True, you're right. It's definitely something we should clean up eventually, but for now let's keep it.

jbarotin committed:
Add ignore error to execute() and use it in get_current_rules()

felixfontein commented:
FYI: `devel` now contains some basic integration tests for `ufw` (see #50374). They currently run only on one of the CI nodes (a VM with RHEL7; I didn't manage to install `ufw` under RHEL8, and all other interesting CI nodes are either FreeBSD, OSX, or running in Docker containers, where `ufw` doesn't work), but that's way better than nothing ;-)

felixfontein commented on lib/ansible/modules/system/ufw.py:
```diff
@@ -303,13 +319,47 @@ def ufw_version():
-            execute(cmd + [['-f'], [states[value]]])
+            if module.check_mode:
+                if value in ['reloaded', 'reset']:
+                    changed = True
```
The module currently doesn't return `changed == True` when called with `state == 'reloaded'` or `state == 'reset'` in non-check-mode. So check-mode should do the same and not set `changed = True` here.

(Except if you fix the behavior for non-check-mode as well :) )

jbarotin replied:
state reloaded and reset should have changed at true. I'll see I can do :)

jbarotin replied:
We have to define the comportment of the `changed` var for reload and reset (with or without check mode) :
- about `reset`, I propose something like this : 
```
if filter_line_that_not_start_with("### tuple", rules_dry) != ""  : 
    changed = True
```
- about `reload`, I don't understand really this function. I propose to keep `changed = True` at each call.

felixfontein replied:
I would even argue that `reset` and `reload` always return `changed == True`, similar to `systemctl reload / restart`.

(I think `reload` simply recreates all ipfilter rules. That's probably useful if you screwed something up when editing them manually, or when you change some configuration, like enabling/disabling IPv6.)

jbarotin replied:
That's simpler so that fine for me :)

felixfontein commented on lib/ansible/modules/system/ufw.py:
```diff
@@ -303,13 +319,47 @@ def ufw_version():
+                    current_level = extract.group(2)
+                    current_on_off_value = extract.group(1)
+                    if value != "off":
+                        if value != "on" and value != current_level:
```
You need a `if current_on_off_value == 'off': changed = True` here (i.e. in the `if` block started here).

felixfontein commented on lib/ansible/modules/system/ufw.py:
```diff
@@ -303,13 +319,47 @@ def ufw_version():
+                    elif current_on_off_value != "off":
+                        changed = True
+                else:
+                    changed = True
```
Also, here, `changed` is only set in check-mode, but *not* in regular non-check-mode. Probably the best is to execute the check-mode code everytime, but in case of non-check-mode, also execute the real command.

felixfontein commented on lib/ansible/modules/system/ufw.py:
```diff
@@ -303,13 +319,47 @@ def ufw_version():
+                else:
+                    changed = True
+            else:
+                execute(cmd + [[command], [value], [params['direction']]])
```
Same problem with check mode here.

felixfontein replied:
Actually, no, this one works.

(I did a mistake while testing; I'm currently adding more tests, you'll see the results soon. They will also contain tests for check mode, even though right now many of them are commented out because they give the wrong result...)

jbarotin replied:
I saw your tests with docker, are they easy to execute ?

felixfontein replied:
Unfortunately `ufw` doesn't seem to work in docker containers, so you'd have to run it on your own machine, or in some VM.

I've put the following playbook (`run-tests-locally.yml`) into `test/integration/targets/ufw/` in my Ansible dev tree; executing it with `-b -K` (so it asks for `sudo` password) runs the test on localhost:
```.yaml
---
# Run with -b -K arguments to ansible-playbook!
- hosts: localhost
  tasks:
  # Run the tests
  - block:
    - include_tasks: tasks/run-test.yml
      with_fileglob:
      - "tasks/tests/*.yml"
    vars:
      role_path: '.'
    always:
    - name: Reset ufw to factory defaults
      ufw:
        state: reset
    - name: Enable ufw (with factory defaults)
      ufw:
        state: enabled
```
(You can also use `ansible-test` to run the tests, but I prefer this lightweight playbook, in particular since the docker feature of `ansible-test` is useless for this test since it doesn't work with docker containers...)

I don't have an important firewall config on my machine, so after `reset` and `enabled` it's back to the state I need. Depending on your firewall config and your network / services running on your machine, you don't want to run this on your local machine... :)

jbarotin replied:
Did you manage to cancel the locale ?  

By the way, I think that `LC_ALL=C` is the way to reset locale to default. ( see https://unix.stackexchange.com/questions/87745/what-does-lc-all-c-do)

jbarotin replied:
In my future push, I added to your shell call :
```
environment:
    LC_ALL: C
```

felixfontein replied:
I'm using `en_US.UTF-8` as my default locale, so I usually don't have problems :) (I gave up using localized OS versions a long time ago, the strange and sometimes half-missing translations kept annoying me too much ;) )

felixfontein commented on lib/ansible/modules/system/ufw.py:
```diff
@@ -303,13 +319,47 @@ def ufw_version():
         if command == 'state':
             states = {'enabled': 'enable', 'disabled': 'disable',
                       'reloaded': 'reload', 'reset': 'reset'}
-            execute(cmd + [['-f'], [states[value]]])
```
I noticed that there seem to be two bugs in `ufw`:
 - `ufw -f reset` does not disable the firewall (as advertised, status is `active` afterwards);
 - `ufw --dry-run -f reset` ignores the `--dry-run` flag (and clears all rules).

felixfontein replied:
(So handling check mode differently will actually improve the module :) )

felixfontein replied:
https://bugs.launchpad.net/ufw/+bug/1810082

jbarotin replied:
Good to know!

felixfontein replied:
BTW, there is another `--dry-run` bug (for `enable` and `disable`): https://bugs.launchpad.net/ufw/+bug/1808727

felixfontein commented:
#50400 got merged today; can you rebase against `devel`? That includes more tests, in particular some about check mode (which currently don't look so good because the current behavior is broken ;) ).

jbarotin committed:
Update after third code review (introduce change in changed status)

jbarotin committed:
Adjust tests and fix some problems (#1)

* 'active' also appears in 'inactive'.

* 'reject' is also a valid option here.

* For example for reloaded, changed will be set back to False here.

* Improve and adjust tests.

jbarotin committed:
Fix after merging integration test

jbarotin committed:
handle "disabled" on default routed

jbarotin commented:
@felixfontein I need access to a docker similar to ci in order to understand why the test not pass. 

Have you any docker command line to give me?

felixfontein commented:
The test in CI does not run in docker, it only runs on a VM. iptables / ufw doesn't work in docker. The only CI instance which runs the `ufw` tests is [RHEL/7.6](https://app.shippable.com/github/ansible/ansible/runs/101956/44/tests).

felixfontein commented:
Interestingly, the test does not fail on my local machine.

jbarotin committed:
Add /var/lib/ufw/.. rules files

jbarotin committed:
add unit test

mattclay commented:
CI failure in integration tests: https://app.shippable.com/github/ansible/ansible/runs/102367/44/tests

felixfontein commented:
Conclusions from experiments in #50848: `ufw --dry-run` only returns IPv4 (IPv6) data when working on IPv4-only (resp. IPv6-only) rules. Since check mode currently compares the `--dry-run` output to all rules, check mode fails.

It doesn't fail for me locally since this only becomes visible when there are some other rules around, which is the case for the RHEL VM. These can be added manually with:
```
ufw allow in 22/tcp
ufw allow in from 224.0.0.251 port 5353 proto udp
ufw allow in from ff02::fb port 5353 proto udp
```
I guess we should modify the tests so that there are always some IPv4 and IPv6 rules around, so that this failure doesn't depend on this specific VM's config.

jbarotin committed:
Fix pep8 formatting error

jbarotin commented:
Yep, I agree too. First I'm working to reproduce this on my own computer, then set up a unit test that fails and to finish fixing it. 

To resume my idea of the fix, we're thinking about checking ipv4 in fields 'from_ip' or 'to_ip'.  If it's the case, we'll filter only ipv4 rules from config files to compare with the `ufw --dry-run` execution. 

Feel free to comment or correct me if I'm wrong.

jbarotin committed:
Separate ipv6 and ipv4 rules process from checkmode

jbarotin committed:
fix non-ascii error on ci

jbarotin commented:
@felixfontein here we go :)

felixfontein commented on lib/ansible/modules/system/ufw.py:
```diff
@@ -207,11 +207,59 @@
+    return filter_line_that_match_func(containsIpv6, content)
+
+
+def containsIpv4(ip):
```
```suggestion
def contains_ipv4(ip):
```

felixfontein replied:
Also, you need a match function (vs. contains). It's probably a good idea to compile the regex expression and store it as a global variable, and then use `compiled_regex.search()` and `compiled_regex.match()` instead.

felixfontein commented on lib/ansible/modules/system/ufw.py:
```diff
@@ -207,11 +207,59 @@
+    return re.search(regexp, ip)
+
+
+def containsIpv6(ip):
```
```suggestion
def contains_ipv6(ip):
```

felixfontein commented on lib/ansible/modules/system/ufw.py:
```diff
@@ -207,11 +207,59 @@
+    return ''.join([line for line in content.splitlines(True) if match_func(line)])
+
+
+# validation pattern provided by :
```
This should be a docstring for the correct function / constant.

felixfontein commented on lib/ansible/modules/system/ufw.py:
```diff
@@ -336,14 +435,48 @@ def ufw_version():
-    (_, post_rules, _) = module.run_command("grep '^### tuple' /lib/ufw/user.rules /lib/ufw/user6.rules /etc/ufw/user.rules /etc/ufw/user6.rules")
-    changed = (pre_state != post_state) or (pre_rules != post_rules)
+            if module.check_mode:
+                # if module.boolean(params['delete']):
```
I think this comment needs some updating.

felixfontein commented on lib/ansible/modules/system/ufw.py:
```diff
@@ -336,14 +435,48 @@ def ufw_version():
+                    keys = ['from_ip', 'to_ip']
+                    for key in keys:
+                        if key not in params:
+                            key.remove(key)
```
```suggestion
                            keys.remove(key)
```
maybe?

Instead of this `for` loop, why not have `keys = [key for key in keys if key in parms]` instead?

Also, both **will** be in `params`. You're more likely interested in the one which isn't `any`?

jbarotin replied:
yep I have removed all this part... it's not usefull in fact.

felixfontein commented on lib/ansible/modules/system/ufw.py:
```diff
@@ -336,14 +435,48 @@ def ufw_version():
+                    if len(keys) > 0:
+                        # ufw dry-run doesn't send all rules so have to compare ipv4 or ipv6 rules
+                        if containsIpv4(params['from_ip']) or containsIpv4(params['to_ip']):
+                            if filter_line_that_contains_ipv4(pre_rules) != rules_dry:
```
You should also filter `rules_dry`, in case ufw's behavior changes.

felixfontein commented on test/integration/targets/ufw/tasks/tests/basic.yml:
```diff
@@ -20,7 +23,7 @@
+    - enable_check is changed
     - enable is changed
     - enable_idem is not changed
     - enable_idem_check is not changed
```
After this line, could you add some random IPv4 and IPv6 rules? That will ensure that we don't have such problems again where it fails on shippable, but succeeds on other machines.

jbarotin replied:
I'd like to do such stuff but I think I spent enough time on this PR and there is already lot of tests added.

felixfontein replied:
No problem, I fully understand that :) This should do the trick:
```suggestion
    - enable_idem_check is not changed

- ufw:
    rule: allow
    port: 24
- ufw:
    rule: deny
    port: 25
```
At least, on my machine, the test suite still works afterwards :-) And you can get this into the PR with one click ;-)

felixfontein commented on test/units/modules/system/test_ufw.py:
```diff
@@ -0,0 +1,254 @@
+Default: deny (incoming), allow (outgoing), deny (routed)
+New profiles: skip
+
+Vers                       Action      De
```
I guess you should adjust this line as well so it's from an untranslated `ufw`.

jbarotin commented:
@felixfontein thanks for the review, I'll change my code considering your remarks (sorry for the "javaist" function naming, It's my current professional language). To be continued :)

jbarotin committed:
Some change after review

felixfontein commented on lib/ansible/modules/system/ufw.py:
```diff
@@ -336,14 +440,34 @@ def ufw_version():
+
+                    rules_dry = filter_line_that_not_start_with("### tuple", rules_dry)
+                    # ufw dry-run doesn't send all rules so have to compare ipv4 or ipv6 rules
+                    if is_match_ipv4(params['from_ip']) or is_match_ipv4(params['to_ip']):
```
Sorry, I think I got this wrong. Using a match function here is a bad idea, since `from_ip` and `to_ip` can also be subnets, not only addresses, and your regexp does not match subnets in the format `<address>/<bitsize>`. So we either have to use contains here, or we have to extend the regexps. Adding something like `(?:/[0-9]+)?` to them should do the job, though.

jbarotin replied:
I think that we can consider after ufw execution with a code return equal to 0 mean that the ip address is valid. 

So we can perform a simple seek of the IP address pattern to check what format the rules is targeting (ipv4, ipv6 or both).

felixfontein replied:
Yep, that's true. Either use `search` instead of `match`, or extend the pattern (or both).

jbarotin replied:
In fact using the match function correspond to ^PATTERN so the current code is working for me in this configuration. As you can see in the unit test I've added.

https://docs.python.org/2/library/re.html#search-vs-match

felixfontein replied:
Interestingly, I always thought that `match()` corresponds to `^PATTERN$` and not just `^PATTERN`. In that case, everything's fine :)

(Except that the function's name could be more precise, `is_match_ipv4()` sounds to me that it only matches a IPv4 address with nothing else after it. But that's a matter of taste, I guess.)

jbarotin replied:
with unit test, renaming is not a problem :)

felixfontein commented on lib/ansible/modules/system/ufw.py:
```diff
@@ -336,14 +440,34 @@ def ufw_version():
+                    if is_match_ipv4(params['from_ip']) or is_match_ipv4(params['to_ip']):
+                        if filter_line_that_contains_ipv4(pre_rules) != filter_line_that_contains_ipv4(rules_dry):
+                            changed = True
+                    elif is_match_ipv6(params['from_ip']) or is_match_ipv6(params['to_ip']):
```
(same here)

jbarotin committed:
Add unit test with sub network mask

jbarotin committed:
rename is_match function by is_starting

felixfontein reviewed (approved):
I think this is ready to go!

shipit

felixfontein commented:
I've added the PR to a list of PRs to be reviewed on the next Monthly Big PR review (will be held on February 21st). Maybe there will already be more reviews before that, but maybe not :)

Are any of the maintainers @ahtik @ovcharenko @pyykkis still around?

jbarotin commented:
Thanks felix, by the way I'm beginning to work on the diff mode (strike while the iron is hot :-)). I have work to do, but when I'll be ready, I'll need this PR merged.

felixfontein commented:
What you can already do is create a new branch based off this branch, which includes this PR and some more commits, and create a PR for it. Just be sure to mark it as `[WIP]` and add a link to this PR, and once this PR is merged to rebase the new PR. Then you can already get the benefits or Ansible's CI and possible feedback from others (like me).

ovcharenko commented:
LGTM.
shipit

ovcharenko commented:
bot_status

felixfontein commented:
bot_status

felixfontein commented:
shipit

felixfontein commented:
Looks like ansibot had some hickups...

felixfontein commented:
Anyway, there's something we forgot: we need a [changelog fragment](https://docs.ansible.com/ansible/latest/reference_appendices/release_and_maintenance.html#changelogs-how-to). @jbarotin can you add one (the type would be `minor_changes:`)?

jbarotin committed:
add changelog fragment

felixfontein commented:
The only change during the last force push / rebase to ufw related files was the addition of the changelog, so I'm still for

shipit

felixfontein commented:
@jbarotin thanks a lot for working on this! I'm glad it finally got merged :) So now let's work on more ufw features ;-)
@ovcharenko thanks for reviewing this!
@gundalow thanks for merging!

felixfontein commented:
@jbarotin @ovcharenko I would be glad if you could also take a look at my ufw PRs; I've rebased them after this PR got merged: #51166 #50402 #49796 (simplest ones first :) )
